Combobox Styles

Fylax opened the pull request:
With this PR I replicated the fluent design style for Combobox.

The style works properly with mouse and multitouch devices.
I haven't tried it with a stylus (nor RTL cultures).

Known issues:
- Keyboard focus produces coloured border but I wasn't able to set the background properly

TODOs:
- FluentWPF/Styles/ComboBox.xaml
  - on line 151there is a reference to `ContentControlThemeFontFamily`. I am not sure if you already implemented it. In case you didn't, [here](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.media.fontfamily.xamlautofontfamily) a reference to proper values

Finally, the `ComboboxItem` style should have an additional `Reveal` style (to apply on `SelectedPressed` visual state). That is beyond my skills.


One final word: I added support for C#8 for the project. Since it already targets .NET Core 3, it has no contraindications with older frameworks.

OniOniOn- commented:
I've successfully built with your combobox files but I can't get style to work.
Is it not supposed to be like this?
`<ComboBox Style="{StaticResource FluentComboBox}">`

Fylax commented:
Hi, I will check it asap.
In the meanwhile I just noticed that I didn't add the combobox style to `Controls.xaml`.
Can you please manually add to your `ResourceDictionary` the `ComboBox.xaml` file as a temp workaround, just to confirm that it's just an import issue?

Fylax commented:
@OniOniOn- I just updated the code. Can you please confirm that it works properly now?

OniOniOn- commented:
![image](https://user-images.githubusercontent.com/6011621/81381373-96fba080-9147-11ea-9fc7-de888578b6a2.png)
For some reason IntelliSense can't find it and I have to write it manually.
Recommendations:
BorderThinkness is not set to 2 and setting it manually also doesn't work.
(when I first wrote {StaticResource FluentComboBox}, VS said something about BorderThinkness value was set wrong but it never showed up again)
Little tweaks like ContentVerticalAlignment="Center"?
Border color when mouse not over should be #FF999999 but is set to mouseover color #FF666666.
While clicking unselected ComboBoxItem, items below it are moved downward, maybe reduce height? by amout of border? increment?
I'm complete newbie to WPF so things might not make sense.
Anyway, thank you for your amazing work and help!

Fylax commented:
There are so many `BorderThickness`.
Can you pleas further explain what are you talking about?

Moreover, talking about colors makes no sense as they change depending on system accent and Light/Dark theme.

I tested it mostly on a dark theme honestly.

Can you create a gif/video showing the differences between this combobox and the one in system apps?

OniOniOn- commented:
1. Setting BorderThinkness do nothing. Expected border to be thicker.
![image](https://user-images.githubusercontent.com/6011621/81385411-476ca300-914e-11ea-8b84-e82aa0d32c82.png)

2. (I'm using light theme) Hover BorderColor is too dark. Expected it to be like control panel.
![image](https://user-images.githubusercontent.com/6011621/81387074-0b870d00-9151-11ea-8db1-e49589755b9e.png)


3. Dropdown popup's top border is missing.
4. Dropdown popup's background should be SystemChromeMediumColor instead of SystemChromeMediumLowColor.
5. Clicking unselected item in dropdown will move below items:
![ezgif-3-5373bbe7b576](https://user-images.githubusercontent.com/6011621/81385885-12ad1b80-914f-11ea-85cc-15e5493cc8c8.gif)

All colors were compared/checked with Windows paint color picker

OniOniOn- commented:
In ComboBox.xaml, line 173, ToggleButton is missing BorderBrush="{TemplateBinding BorderBrush}", and BorderThickness="{TemplateBinding BorderThickness}", which stops it from inheriting border from user code. I added it and ran, but all combobox border change at once.

Fylax commented:
@OniOniOn- Thank you for your feedback.
I haven't addressed the color problem, but thanks to your input I think that the BorderThickness is now working as expected.
I set as default thickness (line 176) the value 1. Feel free to test if it looks like the default one in Windows.

OniOniOn- commented:
If you have multiple ComboBox, and MouseOver on one of them, every ComboBox goes to MouseOver state color. I'll try to fix them. (probably )

OniOniOn- commented:
![image](https://user-images.githubusercontent.com/6011621/83672022-9522e100-a610-11ea-9576-0946bfb47648.png)
I changed colors, wrong margins, and origin of popup.
Only some color adjustments and apply reveal style: Done!
I tried to add drop shadow behind popup but my skills were not enough :D

Fylax commented:
Looks great. I think I can give you permissions to write on my repository, and as such on PR, if you wish.

I was looking at your error list. The fourth still isn't clear to me, as on dark theme it looks correct to me. I'm not sure if I'm seeing it wrong or colours are wrongly mapped. I have to use the colour picker ASAP.

Regarding the fifth, it's most likely a margin changing size (while it should be a padding).

Armin2208 commented:
HI! I also tried out your implementation and it looks really great. Thanks :) 

Adding this line of code worked without adding manually to ResourceDictionary:
`<ComboBox Style="{StaticResource FluentComboBox}">`
IntelliSense just throw an error, but after building everything is fine.

I just have following issues:
1. If I hover above one ComboBox, ALL of them are showing the hover Animation. This is kinda weird.
2. After switching from dark to light, or vice versa the hovering color doesn't update.
Would be nice if you could fix this.

Thanks :)

Fylax commented:
I updated a missing `VisualState` for the `ToggleButton` and I fixed a `BindingError`.

I'll try to address first point ASAP, and I totally agree with your second point.

SlackingVeteran commented:
@Fylax Not sure what's causing this but popup's width is little less than actual width of the Combobox i'd say its less than 1

On the image below you can see border of combobox underneath the popup on the left side. I tried to add margin of -1 to left side and doing so right side border was visible more than left one
![image](https://user-images.githubusercontent.com/5356702/86207671-05078700-bb3d-11ea-8a34-ebed7696ca4c.png)

No changes to your code has been made, its all as is from your commits

Fylax commented:
@SlackingVeteran Did you set any width to the combobox?

Fylax commented:
@OniOniOn- The `ScrollViewer` problem was easier than expected (sorry, I forgot to push the changes earlier).
Basically, the `StackPanel` containing items grows in height endlessly.
I set a `MaxHeight` (according to a visual comparison with Windows comboboxes and now it works perfectly.

I set it to use `ScrollBar` style from `ScrollBar.xaml`. It doesn't show as expected, but it's a different matter at this point.

SlackingVeteran commented:
@Fylax  Yes I had to. Otherwise the combobox would take the entire width of the grid

SlackingVeteran commented:
Below is me setting Height and width manually:
![image](https://user-images.githubusercontent.com/5356702/86388093-2746f500-bc62-11ea-8e00-53f41a96cc52.png)

Below is me setting Height only:
![image](https://user-images.githubusercontent.com/5356702/86388171-50678580-bc62-11ea-8401-4c7abef1a47c.png)

On Above image popup width is small but if you look at the left of popup you can still see the border of combobox

Below image is by wrapping ComboBox in Grid and setting height and width of grid instead of ComboBox:
![image](https://user-images.githubusercontent.com/5356702/86388392-ba802a80-bc62-11ea-9b02-b38d2b8bad8e.png)

Above usage is the only time where you cant see that border but popup width is smaller than the width of the ComboBox

Armin2208 commented:
Really nice work & changes guys! The fade animation is nice, scrollbar works and the ComboBox is now really responsive. 
Only one criticism: the "new" style of the combobox looks a little bit thick.

This is the current style:
![image](https://user-images.githubusercontent.com/32982430/86606966-fbe94080-bfa8-11ea-9fe8-8403ebe3b8e3.png)

And this was the one before:
![image](https://user-images.githubusercontent.com/32982430/86606987-03a8e500-bfa9-11ea-87e2-0f2cbda47aea.png)

It just looked a little bit nicer in preview commits/builds.
Also the design of a deactivated combobox is now the same as it would be activated. So there should be again a difference.

Other than that: nice implementation! I am pretty sure this is now ready for a pull request :)

OniOniOn- commented:
@Armin2208  I couldn't find any deactivated ComboBox in my control panel, can you post a picture of "real" deactivated ComboBox? (how it should look)

OniOniOn- commented:
@SlackingVeteran Can you try adding SnapsToDevicePixels="True" to Border at line 230?
@Armin2208 I couldn't find any deactivated ComboBox in my control panel, can you post a picture of "real" deactivated ComboBox? (how it should look in UWP apps)
@Fylax I tested the code, and the height works perfect.

Armin2208 commented:
> @Armin2208 I couldn't find any deactivated ComboBox in my control panel, can you post a picture of "real" deactivated ComboBox? (how it should look in UWP apps)

I found one! In windows settings for projecting to this PC if you don't have wifi.
Looks like this:
![image](https://user-images.githubusercontent.com/32982430/87971474-b27b1400-cac5-11ea-8426-9f7f73defa2e.png)

And one known issue:
![image](https://user-images.githubusercontent.com/32982430/87971496-bad34f00-cac5-11ea-817f-9655814cc6a7.png)
This happens on some machines, but I sadly can't reproduce this :(

sourcechord commented:
Hi @Fylax, @OniOniOn-, and @Armin2208
Thanks for the PR.
And, I am very sorry for left it without merging it.

I'll merge this PR, and release as FluentWPF 0.9.0-preview1.
I'd like to add enhancement for this feature(like Popup's shadow, and appearance with dark theme and so on).
Then, I'll release it as FluentWPF 0.9.0.

Please wait for a while.
